File: haskell/ex5_4.hs
data ArrayList a = ArrayList
data LinkedList a = LinkedList

class ListLike (l :: * -> *) where
  len :: l a -> Int
  add :: l a -> a -> ()

instance ListLike ArrayList where
  len lst = 0
  add lst elem = ()

instance ListLike LinkedList where
  len lst = 0
  add lst elem = ()

addUntil :: ListLike l => l a -> a -> Int -> ()
addUntil lst v l =
  let _ = len lst in
  add lst v

main = return ()

File: scala/ex5_4.scala
class ArrayList[T]
class LinkedList[T]

trait ListLike[L[_]]:
  def length[T](lst: L[T]): Int
  def add[T](lst: L[T], elem: T): Unit

given ListLike[ArrayList] with
  def length[T](lst: ArrayList[T]): Int = 0
  def add[T](lst: ArrayList[T], elem: T): Unit = ()

given ListLike[LinkedList] with
  def length[T](lst: LinkedList[T]): Int = 0
  def add[T](lst: LinkedList[T], elem: T): Unit = ()

def addUntil[L[_]: ListLike, T](lst: L[T], v: T, len: Int) =
  while summon.length[T](lst) < len do
    summon.add[T](lst, v)
